sycl/doc/extensions/supported/sycl_ext_oneapi_weak_object.asciidoc

```diff
@@ -0,0 +1,332 @@
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 5 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This extension is implemented and fully supported by {dpcpp}.
+
+
+== Overview
+
+SYCL 2020 has a number of classes with common reference semantics, which mean
+that copies of objects of these classes will reference the same objects and
+will be considered the same. As a result, these objects will only be fully
+destroyed when all copies have been destroyed. However, there is currently no
+way to have a reference to an object with these semantics without keeping the
+object alive.
+
+This extension adds the notion of a "weak" SYCL object, represented by the
+`weak_object` class. A weak object holds a reference to a SYCL object with
+common reference semantics of the specified type but will not keep the
+underlying object alive. As such, a weak object can be queries about having
+expired and exposes interfaces for locking the underlying object, creating a
+copy of it if it is still alive.
+
+Additionally this extension adds the notion of owner-based ordering to allow for
+use of SYCL object and their weak variants in order-oriented data structures and
+operations. 
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_WEAK_OBJECT` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== The new `weak_object` class
+
+This extension adds the following class:
+
+[source]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+template <typename SyclObject>
+class weak_object {
+public:
+  using object_type = SyclObject;
+
+  constexpr weak_object() noexcept;
+  weak_object(const SyclObject &SYCLObj) noexcept;
+  weak_object(const weak_object &Other) noexcept;
+  weak_object(weak_object &&Other) noexcept;
+
+  weak_object &operator=(const SyclObject &SYCLObj) noexcept;
+  weak_object &operator=(const weak_object &Other) noexcept;
+  weak_object &operator=(weak_object &&Other) noexcept;
+
+  void reset() noexcept;
+  void swap(weak_object &Other) noexcept;
+
+  bool expired() const noexcept;
+  std::optional<SyclObject> try_lock() const noexcept;
+  SyclObject lock() const noexcept;
```

Comment:
This cannot be `noexcept` since it is defined to throw an exception.

Reply:
My bad! Didn't have it in the implementation either.